src/libmw/src/mmr/IMMR.cpp

```diff
@@ -0,0 +1,42 @@
+#include <mw/mmr/MMR.h>
+
+using namespace mmr;
+
+mw::Hash IMMR::Root() const
+{
+    const uint64_t size = GetNextLeafIdx().GetPosition();
+    if (size == 0) {
+        return ZERO_HASH;
+    }
+
+    // Find the "peaks"
+    std::vector<uint64_t> peakIndices;
+
+    uint64_t peakSize = BitUtil::FillOnesToRight(size);
+    uint64_t numLeft = size;
+    uint64_t sumPrevPeaks = 0;
+    while (peakSize != 0) {
+        if (numLeft >= peakSize) {
+            peakIndices.push_back(sumPrevPeaks + peakSize - 1);
+            sumPrevPeaks += peakSize;
+            numLeft -= peakSize;
+        }
+
+        peakSize >>= 1;
+    }
+
+    assert(numLeft == 0);
+
+    // Bag 'em
+    mw::Hash hash = ZERO_HASH;
+    for (auto iter = peakIndices.crbegin(); iter != peakIndices.crend(); iter++) {
+        mw::Hash peakHash = GetHash(Index::At(*iter));
+        if (hash == ZERO_HASH) {
+            hash = peakHash;
+        } else {
+            hash = Node::CreateParent(Index::At(size), peakHash, hash).GetHash();
```

Comment:
In the Grin documentation, it seems to suggest doing Node::CreateParent(Index::At(size), hash, peakHash) instead, i.e. the previous hash goes on the left. Don't know whether the order really matters or not.

Reply:
Wow, you really are thorough lol. Order doesn't really matter. We'll just make sure we thoroughly document how the MMR hashing is done.